pkg/controller/mcpgateway_controller.go

```diff
@@ -0,0 +1,277 @@
+		"serverCount", len(brokerConfig.Servers))
+	return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+}
+
+func (r *MCPGatewayReconciler) writeAggregatedConfig(ctx context.Context, brokerConfig *config.BrokerConfig) error {
+	writer := NewConfigMapWriter(r.Client, r.Scheme)
+	return writer.WriteAggregatedConfig(ctx, ConfigNamespace, ConfigName, brokerConfig)
+}
+
+func isReady(mcpGateway *mcpv1alpha1.MCPGateway) bool {
+	for _, condition := range mcpGateway.Status.Conditions {
+		if condition.Type == "Ready" && condition.Status == metav1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *MCPGatewayReconciler) discoverServersFromHTTPRoutes(ctx context.Context, mcpGateway *mcpv1alpha1.MCPGateway) ([]ServerInfo, error) {
+	var serverInfos []ServerInfo
+
+	for _, targetRef := range mcpGateway.Spec.TargetRefs {
+		// Validate group and kind
+		if targetRef.Group != "gateway.networking.k8s.io" {
+			return nil, fmt.Errorf("invalid targetRef group %q: only gateway.networking.k8s.io is supported", targetRef.Group)
+		}
+		if targetRef.Kind != "HTTPRoute" {
+			return nil, fmt.Errorf("invalid targetRef kind %q: only HTTPRoute is supported", targetRef.Kind)
+		}
+
+		namespace := mcpGateway.Namespace
+		if targetRef.Namespace != "" && targetRef.Namespace != namespace {
+			return nil, fmt.Errorf("cross-namespace reference to %s/%s not allowed without ReferenceGrant support", targetRef.Namespace, targetRef.Name)
+		}
+
+		httpRoute := &gatewayv1.HTTPRoute{}
+		err := r.Get(ctx, types.NamespacedName{
+			Name:      targetRef.Name,
+			Namespace: namespace,
+		}, httpRoute)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return nil, fmt.Errorf("HTTPRoute %s/%s not found", namespace, targetRef.Name)
+			}
+			return nil, fmt.Errorf("failed to get HTTPRoute %s/%s: %w", namespace, targetRef.Name, err)
+		}
+
+		if len(httpRoute.Spec.Rules) == 0 || len(httpRoute.Spec.Rules[0].BackendRefs) == 0 {
+			return nil, fmt.Errorf("HTTPRoute %s/%s has no backend references", namespace, targetRef.Name)
+		}
+
+		backendRef := httpRoute.Spec.Rules[0].BackendRefs[0]
+		if backendRef.Name == "" {
+			return nil, fmt.Errorf("backend reference has no name")
+		}
+
+		kind := "Service"
+		if backendRef.Kind != nil {
+			kind = string(*backendRef.Kind)
+		}
+
+		if kind != "Service" {
+			return nil, fmt.Errorf("backend reference is not a Service: %s", kind)
+		}
+
+		backendNamespace := namespace
+		if backendRef.Namespace != nil {
+			backendNamespace = string(*backendRef.Namespace)
+		}
+
+		port := int32(80)
+		if backendRef.Port != nil {
+			port = int32(*backendRef.Port)
+		}
+
+		endpoint := fmt.Sprintf("http://%s.%s.svc.cluster.local:%d", backendRef.Name, backendNamespace, port)
+
+		toolPrefix := targetRef.ToolPrefix
+		if toolPrefix == "" {
+			toolPrefix = mcpGateway.Spec.ToolPrefix
+		}
+
+		// Extract hostname from HTTPRoute
+		hostname := ""
+		if len(httpRoute.Spec.Hostnames) > 0 {
```

Comment:
not sure multiple hostnames would ever make sense given the role of this HTTPRoute is simply internally router to an MCP backend. Might be better to just bail if !=1

Reply:
makes sense, fixed